### PR TITLE
Ophyd class for square-one tri-sphere motion system

### PR DESCRIPTION
**Pull request by baljamal**
Ophyd class for square-one tri-sphere motion system

## Motivation and Context
TMO  needs Ophyd class for square-one to be able to do scans rather than doing that manually.

## How Has This Been Tested?
An actual test has been done on TMO LAMP square-one table.
X-axis translation moved using ophyd class and cross-validated with LabVIEW interface.

## Where Has This Been Documented?
some inline comments

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:16]: (Optional) There's a shorthand for these: "self." can be omitted.
```suggestion
    setpoint = FCpt(EpicsSignal, '{prefix}:TARGET:{_axis}', kind='normal')
```

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:17]: I suspect the readback should be "hinted" here, it's the most important component for these scans. This will make it show up in the automatic table/plotting callbacks during bluesky scans.
```suggestion
    readback = FCpt(EpicsSignal, '{self.prefix}:TARGET:{self._axis}:RBV', kind='hinted')
```

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:30]: We should be really careful doing this, for three reasons:
1. We're blocking on two PV connections during `__init__`, and we're doing it 6 times for a maximum total of 12 possible timeouts.
2. We're causing a change to the control system on `__init__`. If a second user opens a session that contains a square 1 device while the first user is doing a scan, depending on the timing we can get very strange behavior.
3. This device will error out on `__init__` if run from anywhere outside of tmo, which may be disruptive.

What exactly is gained from doing this sync? Is there any way we can accomplish this during moves only instead of at initialization?

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:31]: Same as above: does this need to happen now, during `__init__`? This is a 2x timeout + 6 axis = 12 timeouts, potentially. It might be better to do this when the limits are checked or by using a subscription callback.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:43]: You've skipped the `Device.__init__` (`super().__init__(prefix, name=name)`) entirely and implemented everything as an `__init__`-time attribute instead of as a component.

These individual subdevices will work, but SQR1 will not function properly as a top-level device.

You can instead include components in the class definition and skip overriding `__init__` entirely.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:28]: This behaves a bit differently to all other devices because the positional `prefix` argument has been made keyword-only and `axis` is now the first argument instead of `prefix`. I recommend undoing this so that people don't get confused when trying to instantiate this device, and to make it easier to include this device as a component.

```suggestion
    def __init__(self, prefix, axis, **kwargs):
        self._axis = axis  # axis values {X,Y,Z,rX, rY, rZ}
        super().__init__(prefix, **kwargs)
```

**Review by ZLLentz — CHANGES_REQUESTED** [review]
I have some questions and suggestions here, to try to avoid subtle issues/confusions in the future. 

It looks like it is good and functional for this use case right now but I want to make sure we'll be OK later too.

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py:30]: This system has one move command PV for the six motors, therefore I initialize each motor setpoint value with its readback value to avoid moving one motor while other motor(s) setpoints are zeros for example, then they will move to unwanted positions.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:32]: Optional suggestion: run this through `black` to see what kind of formatting the python community is favoring for arg inputs these days. This sort of layout is somewhat hard to read.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:33]: Optional suggestion: This is ok but is actually redundant because it is set during `super().__init__`, you could remove it

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:83]: Question: are you finding this print statement useful? If you are, great, if not, you could remove it.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:132]: Does this actually work smoothly with no hiccups? We're writing to the shared actuate signal 6 times in the code here.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:140]: There's funny syntax here that's equivalent and fairly clear/compact
```
status = x_status & y_status & z_status & rx_status & ry_status & rz_status
```

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:141]: This makes it so we sometimes return `None` and sometimes return a status object. Usually it's best to always return the same type when possible.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:56]: Optional suggestion: this isn't wrong, but you don't need to use an FCpt here. Usually FCpt is for when you need to put non-prefix things into the prefix.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:80]: I think this and the other associated code is a clean solution to the setpoint synchronization problem :+1:

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py]: I tried black to auto-format it basically, it put the whole method definition in one line which violates PEP8 80 characters rule

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py]: you are right, I deleted it

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py:121]: these PVs are living in the same CA server, so writing set point values to them is not a heavy process

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py]: I agree, operator overloading looks more neat. I followed your suggestion

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py]: I followed your suggestion

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py:69]: thanks for compliment

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [reply pcdsdevices/sqr1.py:132]: I guess I was more curious about how the IOC behaves. The sequence here is something like:

- call x move:
  - write to the x setpoint
  - set the actuate signal
- call y move:
  - write to the y setpoint
  - set the actuate signal
... etc.

Does this encounter any problems with e.g. starting a move to x before the rz parameter is available?

Comment (ZLLentz) [reply pcdsdevices/sqr1.py:132]: I think this is the last functional thing to consider before merging

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:1]: Aside from the sequencing question, this PR is missing the standard procedure of including release notes.
cd docs -> run `./pre-release-notes.sh 1123 square-one` -> put your username in and say you added the square one device

If the pre-merge checklist from the description template is completed we can merge soon

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py:132]: I tried to move three motors at the same time and it is working fine. I got your point but it doesn't seem a problem.

File: pcdsdevices/sqr1.py
Comment (baljamal) [reply pcdsdevices/sqr1.py]: you are right, I should pay more attention to types.
I went with this: sync_setpoint: Callable | None

Comment (ZLLentz) [reply pcdsdevices/sqr1.py]: This is done elsewhere in the file too

File: docs/source/upcoming_release_notes/1123-square-one_tri-sphere_motion_system.rst
Comment (ZLLentz) [inline docs/source/upcoming_release_notes/1123-square-one_tri-sphere_motion_system.rst:18]: Can you please include context information here? Even if it's just the header of your module docstring?

At this point in the review I'm kind of just saying "read and follow the pre-merge checklist" repeatedly which is probably unpleasant for both of us.

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [reply pcdsdevices/sqr1.py:132]: Alright, if you've tested the code as-is and there hasn't been a problem it's probably OK and shouldn't hold up the PR
I'm still worried about if of course since it's an unnecessary risk

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:5]: Your code doesn't work any more, needs to add
```
from __future__ import annotations
```
at the top here to allow the annotation | syntax in py39

Comment (ZLLentz) [reply pcdsdevices/sqr1.py:5]: A small unit test would catch things like this, even one that simply imports the module

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:50]: These are all malformed now:
```
In [1]: import pcdsdevices.sqr1

In [2]: dev = pcdsdevices.sqr1.SQR1("MCQ:TMO:01", name="dev")

In [3]: dev.component_names
Out[3]: ('x', 'y', 'z', 'rx', 'ry', 'rz', 'stop_signal')

In [4]: dev.x.prefix
Out[4]: 'MCQ:TMO:01{prefix}'
```
the curly braces setup only works for FCpt, to refactor to Cpt you would have needed to drop this down to (I think):
```suggestion
    x = Cpt(SQR1Axis, '', axis="X", rtol=0.001)
```

File: docs/source/upcoming_release_notes/1123-square-one_tri-sphere_motion_system.rst
Comment (ZLLentz) [inline docs/source/upcoming_release_notes/1123-square-one_tri-sphere_motion_system.rst:30]:
```suggestion
- baljamal
```

File: pcdsdevices/sqr1.py
Comment (tangkong) [inline pcdsdevices/sqr1.py:2]: For the record you shouldn't just ignore line length conventions because you feel like it.  It's perfectly possible to break this up into two lines

**Review by ZLLentz — COMMENTED** [review]
I like this as-is and want to merge it
A simple import/instantiation test would be useful but I'm not going to force you to do it

One more thing I need to check before the approval: I want to see how these look in the generated docs, the docstrings look kind of nonstandard but it shouldn't necessarily be a problem

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:185]: These examples render really well in the html docs, we should do this more often :+1:
image

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:141]: Each of these is rendering kind of strangely in the html docs:
image

Comment (ZLLentz) [reply pcdsdevices/sqr1.py:141]: I'm not sure the right way to reformat this because I've only used the numpy style docstrings
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/#google-vs-numpy

Comment (baljamal) [reply pcdsdevices/sqr1.py:141]: I think the problem with second line indentation

File: pcdsdevices/sqr1.py
Comment (ZLLentz) [inline pcdsdevices/sqr1.py:66]: Your indentation commit below fixes the docs build for the SQR1 class
If you do the same thing for the docstring here in the SQRT1Axis then this is ready to merge

File: pcdsdevices/sqr1.py
Comment (klauer) [inline pcdsdevices/sqr1.py:64]: If you look at our style guide [here](https://pcdshub.github.io/style.html#docstrings), or any of the other code in this repository (or any of our other Python repositories), you'll see we follow the numpy docstring convention.

I think it would be good for you to understand how to format them for your own future reference/submissions, but we can reformat them for you if necessary to get this merged sooner rather than later.

Comment (baljamal) [reply pcdsdevices/sqr1.py]: thanks for sharing the link, I wasn't aware of it.

Comment (klauer) [reply pcdsdevices/sqr1.py]: Thanks for fixing it up 👍

**Review by ZLLentz — APPROVED** [review]
Let's finally merge this

**Pull request merged**
